### PR TITLE
feat(eth-sparse-mpt): add ~3x faster v3 state-root implementation and harness tools

### PR DESCRIPTION
**Pull request by bertmiller**
## 📝 Summary

I will state upfront that this code is almost entirely written with Codex. However, the results seem promising and it shares much overlap with the existing v2 state root anyway.

I asked Codex to setup a harness that measured performance across different depths of cache warmth and compared state root implementations to each other and Reth's for correctness. Then I asked it to optimize v2 and implement a v3 state root. This was effective because the agent could simply iterate if it noticed it had the wrong root or if it was slower. Finally, I ran the harness for 1000+ blocks and the new v3 always produced the same root as v2 and reth, so I have some confidence in the implementation. You can find the testing scripts in this PR, specifically `correct_and_bench.rs`. Note that these require unwinding a reth node block by block (thank you Vitaly for this trick).

The results are a significant speedup relative to v2. Over 1,000 blocks:

| Scope | v2 p50 | v3 p50 | v2 p99 | v3 p99 | v2 mean | v3 mean |
  |---|---:|---:|---:|---:|---:|---:|
  | Overall (excluding warm=100, n=3213) | 3.01 | 0.88 | 4.71 | 1.87 | 3.19 | 0.97 |

| Warm % | v2 p50 | v3 p50 | p50 speedup | v2 p99 | v3 p99 | p99 speedup | v2 mean | v3 mean | mean speedup |
  |---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
  | 70 | 4.16 | 1.09 | 3.82x | 5.96 | 2.22 | 2.68x | 4.32 | 1.21 | 3.58x |
  | 80 | 3.05 | 0.94 | 3.26x | 4.75 | 1.93 | 2.46x | 3.23 | 1.03 | 3.14x |
  | 90 | 1.83 | 0.61 | 3.00x | 3.41 | 1.45 | 2.35x | 2.00 | 0.68 | 2.95x |

Here's a moving average:
image

The one notable regression is that in ~2.5% of blocks there were v3 p99 cases that were greater than v2. On average these regressions added 1.89ms, and the worst observed was 23.9ms for a single block.

## 💡 Motivation and Context
Calculating the state root is in the hot path for a new block so making it faster lowers latency for all our users.

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [x] Run `make test`
* [x] Added tests (if applicable)

### CODE REVIEW

**Comment by bertmiller** [comment]
@dvush if you would like me to address Claude's input I'll pass it to Codex, let me know either way.

File: crates/eth-sparse-mpt/src/v_experimental/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1217 @@
...
+            }
+            let storage_calc = self.get_account_storage(address);
+            let mut storage_calc = storage_calc.lock();
+            let storage_calc = &mut *storage_calc;
+            for missing_node in storage_calc.missing_nodes_requested.drain(..) {
+                if storage_calc.proof_store.has_proof(&missing_node) {
+                    let ok = storage_calc.trie.try_add_proof_from_proof_store(&missing_node, &storage_calc.proof_store).expect("should be able to insert proofs from proof store when they are found (storage trie)");
+                    assert!(ok, "proof is not added (storage trie)");
+                } else {
+                    panic!("Missing node that was just fetched is not there (storage trie)");
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/mod.rs:671]: Same behavior as v2

File: crates/eth-sparse-mpt/src/v_experimental/trie/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1211 @@
...
+            let node = self
+                .nodes
+                .get(current_node)
+                .ok_or_else(|| NodeNotFound::new(ins_key))?;
+            self.hashed_nodes[current_node] = false;
+            match node {
+                DiffTrieNode::Branch { children } => {
+                    let children = *children;
+
+                    let n = ins_key[path_walked] as usize;
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/trie/mod.rs:239]: Same behavior as v2

File: crates/eth-sparse-mpt/src/v_experimental/trie/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1211 @@
...
+    // 3. root_hash ensures that we don't have a race here
+    fn calculate_rlp_pointer_node(
+        &self,
+        node_idx: usize,
+        rlp: &mut Vec<u8>,
+        proof_store: &ProofStore,
+    ) {
+        self.rlp_encode_node(node_idx, rlp, proof_store);
+        let result =
+            unsafe { &mut *(self.rlp_ptrs_local.as_ptr().add(node_idx) as *mut ArrayVec<u8, 33>) };
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/trie/mod.rs:788]: Same behavior as v2

File: crates/eth-sparse-mpt/src/v_experimental/trie/proof_store.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    pub fn has_proof(&self, key: &Nibbles) -> bool {
+        self.proofs.contains_key(key)
+    }
+
+    pub fn add_proof<P: AsRef<[u8]>>(
+        &self,
+        key: Nibbles,
+        proof: Vec<(Nibbles, P)>,
+    ) -> Result<(), alloy_rlp::Error> {
+        if self.proofs.contains_key(&key) {
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/trie/proof_store.rs:59]: Same behavior as v2

File: crates/eth-sparse-mpt/src/v_experimental/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1217 @@
...
+                    &StorageTrieStatus::InsertsProcessed,
+                    "storage trie is not updated properly"
+                );
+                let storage_calc = self.get_account_storage(address);
+                let mut storage_calc = storage_calc.lock();
+                let storage_calc = &mut *storage_calc;
+                let hash = storage_calc
+                    .trie
+                    .root_hash(PARALLEL_HASHING_STORAGE_NODES, &storage_calc.proof_store)
+                    .expect("missing node while hahsing storage trie");
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/mod.rs:697]: Same as v2, but happy to fix if we want

File: crates/eth-sparse-mpt/src/v_experimental/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1217 @@
...
+        for (address, proof) in self.account_trie.proof_result.drain(..) {
+            proofs.insert(
+                address,
+                proof.into_iter().map(|(_, node)| node.into()).collect(),
+            );
+        }
+        for proof_target in proof_targets {
+            if !proofs.contains_key(proof_target) {
+                return Err(SparseTrieError::Other(eyre::eyre!(
+                    "Proof was not fethed correctly"
```
[the remark this answers is not in the document]
Comment (bertmiller) [reply crates/eth-sparse-mpt/src/v_experimental/mod.rs:1076]: Same as v2, but happy to fix if we want

**Comment by bertmiller** [comment]
I think the remaining Claude comments are things that were copied over from v2 (except the small changes in v2, which I can revert if you want. not sure why it did that)

**Comment by dvush** [comment]
@bertmiller 

Interesting work, I think this can be used to improve this code or something like that where you can define objective benchmark.

Also, ignore what claude reviewer says because it complains essentially about v2 version.

I checked v3 implementation and its a comparatively small diff compared to v2 which is good. 

My first impressions is that its probably "cheating" a bit and its likely not a 3x improvement in practice, although it might be. My local reth broke so I need to resync to quickly verify.

1. there are microbenchmarks for the tries (trie_do_bench) where inserting and hashing is verified
`cargo bench --bench trie_do_bench` when I replace v2 with v3 I don't see any difference which is suspicious and I would expect to see some difference for x3 improvement.
2. The one place in a diff that caught my attention is that v3 spawns tasks only for the top level node. So it parallelize only children of the top node. v2 also creates tasks for subtries. I remember that being a problem sometimes on a machine with a lot of cpus but I also remember that its mitigated if you limit number of threads in rayon threadpool to something like 4 or 8 which you would want to do anyway. 

Do you test on the machine with a lot of cores?
If so can you try running your tests with limited number of threads.

**Comment by bertmiller** [comment]
@dvush Yeah I think it's an exciting direction for the use of agents because they'll just keep iterating till they get it right. And they are quite good at figuring out what the intermediate steps should be too.

This is my machine's specs:
- CPU: AMD Ryzen 7 9700X
- Cores/Threads: 8 physical cores, 16 logical CPUs (1 socket, 2 threads/core)
  - RAM: 60 GiB total (16 GiB used, 1.4 GiB free, 44 GiB available)
  - Disk /: 868 GiB total (433 GiB used, 392 GiB free, 53% used)
  - OS: Debian GNU/Linux 13 (trixie)
  - Kernel: 6.12.63+deb13-amd64
  - Arch: x86_64
  

  
 I ran a quick 25 block test (warm 70/80/90), these are the results:
  | Threads | v2 p50 (ms) | v3 p50 (ms) | p50 v2/v3 | v2 mean (ms) | v3 mean (ms) | mean v2/v3 | v2 p99 (ms) | v3 p99 (ms) | p99 v2/v3 |
  |---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
  | 1 | 2.812 | 2.203 | 1.277x | 2.818 | 2.202 | 1.280x | 3.252 | 2.601 | 1.250x |
  | 2 | 1.925 | 1.152 | 1.671x | 1.960 | 1.182 | 1.658x | 2.332 | 1.455 | 1.603x |
  | 4 | 2.135 | 0.777 | 2.749x | 2.150 | 0.805 | 2.672x | 2.375 | 1.009 | 2.354x |
  | 8 | 2.557 | 0.628 | 4.069x | 2.568 | 0.635 | 4.047x | 2.776 | 0.786 | 3.530x |


Script:
```

#!/usr/bin/env bash
set -euo pipefail

# Usage:
#   scripts/run_correct_and_bench_thread_sweep.sh [OUTDIR]
#
# Optional environment overrides:
#   THREADS="1 2 4 8"
#   BIN=/home/robert/rbuilder/target/release/correct_and_bench
#   RETH_BIN=/usr/local/bin/reth
#   DATADIR=/mnt/reth-full
#   CHAIN=mainnet
#   BLOCKS=25
#   CACHE_WARM_ITERS=25
#   CACHE_WARM_KEYS=10000
#   CACHE_WARM_PERCENTAGES=70,80,90,100
#   STATIC_FILE_BLOCKS_PER_FILE=10000

THREADS="${THREADS:-1 2 4 8}"
BIN="${BIN:-/home/robert/rbuilder/target/release/correct_and_bench}"
RETH_BIN="${RETH_BIN:-/usr/local/bin/reth}"
DATADIR="${DATADIR:-/mnt/reth-full}"
CHAIN="${CHAIN:-mainnet}"
BLOCKS="${BLOCKS:-25}"
CACHE_WARM_ITERS="${CACHE_WARM_ITERS:-25}"
CACHE_WARM_KEYS="${CACHE_WARM_KEYS:-10000}"
CACHE_WARM_PERCENTAGES="${CACHE_WARM_PERCENTAGES:-70,80,90,100}"
# /mnt/reth-full commonly uses 10k static-file windows near the tip.
STATIC_FILE_BLOCKS_PER_FILE="${STATIC_FILE_BLOCKS_PER_FILE:-10000}"

OUTDIR="${1:-${OUTDIR:-bench-results/correct-and-bench-mainnet-${BLOCKS}blocks-$(date +%F-%H%M%S)}}"
mkdir -p "$OUTDIR"
echo "output dir: $OUTDIR"

for t in $THREADS; do
  echo "running threads=$t ..."
  args=(
    --cache-warm-iters "$CACHE_WARM_ITERS"
    --cache-warm-keys "$CACHE_WARM_KEYS"
    --cache-warm-percentages "$CACHE_WARM_PERCENTAGES"
    --cache-warm-out-csv "$OUTDIR/correct-and-bench-threads-$t.csv"
    --full
    --blocks "$BLOCKS"
    --reth-bin "$RETH_BIN"
    --datadir "$DATADIR"
    --chain "$CHAIN"
  )
  if [[ -n "$STATIC_FILE_BLOCKS_PER_FILE" ]]; then
    args+=(--static-file-blocks-per-file "$STATIC_FILE_BLOCKS_PER_FILE")
  fi

  if ! RAYON_NUM_THREADS="$t" "$BIN" "${args[@]}" \
    > "$OUTDIR/correct-and-bench-threads-$t.log" 2>&1; then
    echo "failed for threads=$t"
    echo "log: $OUTDIR/correct-and-bench-threads-$t.log"
    tail -n 80 "$OUTDIR/correct-and-bench-threads-$t.log" || true
    exit 1
  fi
  echo "completed threads=$t"
done

{
  echo "threads,iteration,v3_root_correct,warm_pct,v2_p50_ms,v3_p50_ms,v2_p99_ms,v3_p99_ms,v2_mean_ms,v3_mean_ms,speedup"
  for t in $THREADS; do
    tail -n +2 "$OUTDIR/correct-and-bench-threads-$t.csv" | sed "s/^/$t,/"
  done
} > "$OUTDIR/correct-and-bench-thread-sweep-combined.csv"

echo "done: $OUTDIR"
```

**Comment by dvush** [comment]
@bertmiller do you see any difference in `cargo bench --bench trie_do_bench -- insert_nodes_do_trie` when you change v2 vs v3 on your machine? I tried 10k nodes and 100k nodes and I don't see any difference in time to hash it in the case of empty trie and trie with proofs. 

You can create a separate benchmark based on criterion that will just compare v2 vs v3 and there the first thing to test is just empty tries where you insert N element.

The other thing to check is with proofs (which might be the case where v3 wins). gather_and_hash_do_trie bench has example of how to do more real case where we add proofs into spare trie and modify, maybe not exactly this benchmark is needed but something based on that.

I think criterion benches are easier because they are very isolated and specific and its much tighter loop for agent to optimize then based on the whole blocks.

But results on the real blocks are interesting, maybe I miss something here, we need to deploy on staging and see how metrics change. 

Said that, I wonder how we can merge this. 
My thinking is that having a copy of the trie for agent to work with is nice but its hard to fully review and use.
What we can do with this set of changes:
* have v3 merged (maybe call is experimental or something like that) 
* this experimental version is not supposed to be used other than for discovering a better approach
* we take things from v3 and add to v2
 
This will work unless v3 is not a fundamentally different approach which I think might happen.

Tests as sole acceptance criterion are kind of tricky, I had a case where trie was breaking once every other day because of the weird edge case that would trigger if you change account A that was a trie neighbor of the account B that you would later change in the same block. This is not reproducible even if you test block by block. 

What are your thoughts on this?

File: crates/eth-sparse-mpt/Cargo.toml [hunk 7]
```diff
@@ -22,11 +22,19 @@ nybbles = { version = "0.3.3", features = ["serde"] }
 tracing.workspace = true
 
 # reth
+reth-chainspec = { workspace = true, optional = true }
```
Comment (dvush) [inline crates/eth-sparse-mpt/Cargo.toml:25]: can this be a dev dependencies?

**Comment by bertmiller** [comment]
@dvush I like the idea of adding it under experimental and letting it run on staging for awhile. I also like the idea of having an experimental version that agents can change as much as they want, and hand picking ideas from that to the core implementation. It makes me nervous to have a state root implementation that someone doesn't deeply understand.

On the other hand, though, does it really matter if someone understands deeply if it produces the correct result (measured by tests, time in staging, etc)? Perhaps the focus should be on getting the right environment for testing such that you're confident in the code even if you haven't written it yourself. If we need to cherry pick all the stuff that we want to implement then it'll slow down velocity and that'll compound over other domains too (imagine doing this same exercise but for building or whatever else).

 These are my v2 and v3 bench results that you requested:
 
 ```
 v2
   Benchmarking insert_nodes_do_trie: Warming up for 3.0000 s
  Benchmarking insert_nodes_do_trie: Collecting 100 samples in estimated 5.0960 s (1700 iterations)
  Benchmarking insert_nodes_do_trie: Analyzing
  insert_nodes_do_trie    time:   [2.9948 ms 3.0150 ms 3.0359 ms]
  
 v3
  Benchmarking insert_nodes_do_trie: Warming up for 3.0000 s
  Benchmarking insert_nodes_do_trie: Collecting 100 samples in estimated 9.8878 s (10k iterations)
  Benchmarking insert_nodes_do_trie: Analyzing
  insert_nodes_do_trie    time:   [988.20 µs 999.85 µs 1.0125 ms]
                          change: [-67.692% -67.294% -66.888%] (p = 0.00 < 0.05)```

**Comment by dvush** [comment]
@bertmiller I tries to figure why I don't  see any difference so I run it on multiple machines and here is what I found.

* On my machine there is no difference at all (with or without rayon num threads, parallel vs non-parallel), big or small trie.
* But I tried on our staging and on my macbook and its indeed faster and I saw 1.4x(on macbook), 3.8x (on staging)  in microbenchmark so it really works.

This is really good find, because I was always confused why when its deployed its so much worse then it should be. I have a "special" config for some reason and this rayon overhead does not affect it at all.

The differences go away when I do parallelization only on the top level like v3 does https://github.com/flashbots/rbuilder/commit/5d87a4e75f087fccc4c6c079bc428f21f2e5c9ee 

This is a couple of lines change to v2. Can you add it to v2 here please? Also revert current diff to v2 (its seems unnecessary). Then we can merge this and it would have immediate impact.

**Comment by dvush** [comment]
>On the other hand, though, does it really matter if someone understands deeply if it produces the correct result (measured by tests, time in staging, etc)? Perhaps the focus should be on getting the right environment for testing such that you're confident in the code even if you haven't written it yourself. If we need to cherry pick all the stuff that we want to implement then it'll slow down velocity and that'll compound over other domains too (imagine doing this same exercise but for building or whatever else).

I don't mind code not written by myself obviously its just a practical consideration. If changes are small diffs to existing code its much faster to approve and deploy so it would be better to do it this way if its possible. Like here we can fix rayon bug in v2 and be sure that its correct.

If the code is created by AI that was running autonomously for a while and its a multi thousand line change then its a different thing. I don't think its worth it to commit AI code that you don't understand.

Like for this specific problem of tries you can do it more blindly for submodule `trie/` because its pure isolated piece and you can easily test every branch of the algorithm and do fuzz / random tests. 

But when you go to the part where it connects everything: reth, fetching, incremental proofs its more dangerous. Its fragile and there are no good tests for it yet.

We also don't have a good way to debug if anything happens live since its very state-full process if we had more debug infra around it it would also be easier to do it more blindly: 
do unit tests -> do reth based tests with reverts for 1kblocks -> deploy to staging for some time -> deploy live -> if error we dump data and let ai debug it

**Comment by bertmiller** [comment]
>I don't mind code not written by myself obviously its just a practical consideration. If changes are small diffs to existing code its much faster to approve and deploy so it would be better to do it this way if its possible. Like here we can fix rayon bug in v2 and be sure that its correct.

sorry I misunderstood what you were thinking about before, I've been thinking too much about agents :)



> @bertmiller I tries to figure why I don't see any difference so I run it on multiple machines and here is what I found.
> 
> * On my machine there is no difference at all (with or without rayon num threads, parallel vs non-parallel), big or small trie.
> * But I tried on our staging and on my macbook and its indeed faster and I saw 1.4x(on macbook), 3.8x (on staging)  in microbenchmark so it really works.
> 
> This is really good find, because I was always confused why when its deployed its so much worse then it should be. I have a "special" config for some reason and this rayon overhead does not affect it at all.
> 
> The differences go away when I do parallelization only on the top level like v3 does [5d87a4e](https://github.com/flashbots/rbuilder/commit/5d87a4e75f087fccc4c6c079bc428f21f2e5c9ee)
> 
> This is a couple of lines change to v2. Can you add it to v2 here please? Also revert current diff to v2 (its seems unnecessary). Then we can merge this and it would have immediate impact.

Great investigation. I will add this change to v2, revert current v2 diff, and remove v3 so we can merge. Good outcome!

**Comment by dvush** [comment]
@bertmiller I thought that we can commit v3 if it's useful for future work.

**Comment by bertmiller** [comment]
@dvush sounds good. I'll label it "experimental" or something so someone doesn't accidentally enable it because v3 > v2

**Comment by bertmiller** [comment]
@dvush ok, I picked the diff on v2 from the commit you linked, applied it, renamed v3 to v_experimental (long but whatever?). I chose to not address claude's comments to keep the diff between v2 and experimental as small as possible.

Then I found something else which is on experimental. I added a [custom Clone impl](https://github.com/flashbots/rbuilder/commit/305def8ce59b4980b53505ce02e5adc2cae49a8a) that preserves vector capacity (instead of derive clone). A quick test gave me between 5% and ~40% speedup relative to v2, depending on how warm the cache is. It's another focused changed so if you agree we can carry it over to v2 also.

**Review by dvush — APPROVED** [review]
@bertmiller lets merge this and do the rest separately

**Comment by bertmiller** [comment]
Sounds good.
